test-suite/bonds.cpp

```diff
@@ -1378,6 +1378,49 @@ void BondTest::testBondFromScheduleWithDateVector()
     }
 }
 
+void BondTest::testThiry360BondWithSettlementOn31st(){
+    BOOST_TEST_MESSAGE("Testing Thiry360Bond With Settlement On 31st of the month");
+
+    // cusip 3130A0X70, data is from Bloomberg
+    SavedSettings backup;
+    Settings::instance().evaluationDate() = Date(28, July, 2017);
+
+    Date datedDate(13, February, 2014);
+    Date settlement(31, July, 2017);
+    Date maturity(13, August, 2018);
+
+    DayCounter dayCounter = Thirty360(Thirty360::USA);
+    Compounding compounding = Compounded;
+
+    Schedule fixedBondSchedule(datedDate,
+            maturity,
+            Period(Semiannual),
+            UnitedStates(UnitedStates::GovernmentBond),
+            Unadjusted, Unadjusted, DateGeneration::Forward, false);
+
+    FixedRateBond fixedRateBond(
+            1,
+            100,
+            fixedBondSchedule,
+            std::vector<Rate>(1, 0.015),
+            dayCounter,
+            Unadjusted,
+            100.0);
+
+    double cleanPrice = 100;
+
+    Real yield = BondFunctions::yield(fixedRateBond, cleanPrice, dayCounter, compounding, Semiannual, settlement);
+    ASSERT_CLOSE("yield", settlement, yield, 0.015, 1e-3);
```

Comment:
1e-3 is an absolute tolerance, which means that a calculated yield of 0.0146 would pass the test.  I'd go for a smaller one.